src/java/org/apache/cassandra/streaming/async/NettyStreamingMessageSender.java

```diff
@@ -0,0 +1,354 @@
+
+    private final StreamSession session;
+    private final int protocolVersion;
+    private final OutboundConnectionIdentifier connectionId;
+    private final StreamConnectionFactory factory;
+
+    private volatile boolean closed;
+    private Channel controlMessageChannel;
+
+    // note: this really doesn't need to be a LBQ, just something that's thread safe
+    private Collection<ScheduledFuture<?>> channelKeepAlives = new LinkedBlockingQueue<>();
+
+    private final ThreadPoolExecutor fileTransferExecutor;
+
+    public NettyStreamingMessageSender(StreamSession session, OutboundConnectionIdentifier connectionId, StreamConnectionFactory factory, int protocolVersion)
+    {
+        this.session = session;
+        this.protocolVersion = protocolVersion;
+        this.connectionId = connectionId;
+        this.factory = factory;
+
+        String name = session.peer.toString().replace(':', '.');
+        fileTransferExecutor = new DebuggableThreadPoolExecutor(1, DEFAULT_MAX_PARALLEL_TRANSFERS, 1L, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>(),
+                                                                new NamedThreadFactory("NettyStreaming-Outbound-" + name));
+        fileTransferExecutor.allowCoreThreadTimeOut(true);
+    }
+
+    @Override
+    public void initialize() throws IOException
+    {
+        StreamInitMessage message = new StreamInitMessage(FBUtilities.getBroadcastAddress(),
+                                                                    session.sessionIndex(),
+                                                                    session.planId(),
+                                                                    session.description(),
+                                                                    session.keepSSTableLevel(),
+                                                                    session.isIncremental(),
+                                                                    session.getPendingRepair());
+        sendMessage(message);
+    }
+
+    public boolean hasControlChannel()
+    {
+        return controlMessageChannel != null;
+    }
+
+    public void injectControlMessageChannel(Channel channel)
+    {
+        this.controlMessageChannel = channel;
+    }
+
+    private void setupControlMessageChannel() throws IOException
+    {
+        if (controlMessageChannel == null)
+            controlMessageChannel = createChannel();
+    }
+
+    private Channel createChannel() throws IOException
+    {
+        Channel channel = factory.createConnection(connectionId, protocolVersion);
+        ChannelPipeline pipeline = channel.pipeline();
+        pipeline.addLast(NettyFactory.instance.streamingGroup, NettyFactory.INBOUND_STREAM_HANDLER_NAME, new StreamingInboundHandler(connectionId.remoteAddress(), protocolVersion, session));
+
+        int keepAlivePeriod = DatabaseDescriptor.getStreamingKeepAlivePeriod();
```

Comment:
> `DatabaseDescriptor.getStreamingSocketTimeout()`?

I believe that is a left over from CASSANDRA-11841 and was deprecated with 3.10. I'll remove for 4.0.